### PR TITLE
[juju] Support for Juju version 2

### PR DESCRIPTION
**Pull request by nniehoff**
Adding support for Juju 2.x.  This plugin needed some cleanup.  There are 3
use cases for the juju plugin: the juju client, a juju deployed machine, and
a juju controller.  I have broken the plugin into 3 parts to reflect this.

The juju plugin is now specifically for the juju client use case.  Juju is
typically installed as a snap or a package, therefore I added /snap/bin to the
path for Debian to handle this until a better method for handling snaps is
figured out.  Juju is configured as a user to communicate with the controller
it is therefore necessary to run the juju commands with sudo for them to
execute properly.  This plugin now collects as much information about the juju
environment as possible by using the juju command, this information includes
the configured clouds, models, spaces, model configs, status of each model,
and configuration of each app in each model.

The juju_machine plugin is for the machines deployed by juju use case.  This
plugin handles config, log and journal collection for a juju machine.  There
is no package installed to indicate it is a juju machine, so this plugin is
activated by the presence of /var/log/juju.

The juju_controller plugin is for the juju controller use case.  A juju
controller is a juju machine so the machine plugin should handle the
collection of logs from that aspect.  The only identifying piece of
information that this is a juju controller is the /var/lib/juju/db directory
and really juju is just a mongodb server.  This plugin instruments some mongo
queries to get the status of that database.

Signed-off-by: Nick Niehoff nick.niehoff@canonical.com

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X] Is the subject and message clear and concise?
- [X] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [X] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?

### CODE REVIEW

File: sos/plugins/juju.py [hunk 1]
```diff
@@ -9,113 +9,201 @@
...
+    plugin_name = "juju"
+    profiles = ("virt", "sysmgmt")
+    files = ("/snap/bin/juju", "/usr/bin/juju")
 
     option_list = [
-        ('export-mongodb',
-         'Export mongodb collections as json files', '', False),
-        ('generate-bundle',
-         """Generate a YAML bundle of the current environment
-         (requires juju-deployerizer)""", '', False),
```
Comment (bmr-cymru) [inline sos/plugins/juju.py]: Note that if you are deleting plugin options, then any automation that tries to use them will break with the updated plugin. We tend to avoid this unless absolutely necessary for this reason.

File: sos/plugins/juju.py [hunk 2]
```diff
@@ -9,113 +9,201 @@
...
+                    ]
+                    for cmd in model_cmds:
+                        # The file names can be long and confusing from the
+                        # juju commands, clean them up and give them an
+                        # appropriate extension if possible
+                        filename = self._mangle_command(cmd)
+                        if re.match(r".*_--format_yaml.*", filename):
+                            filename = re.sub(r"_--format_yaml", "", filename)
+                            filename += ".yaml"
+                        self.add_cmd_output(
```
Comment (bmr-cymru) [inline sos/plugins/juju.py]: What's the point of all this complexity? Why is it necessary or useful to strip out `_--format_yaml`? This seems like a very odd way to treat the command file names.

Comment (bmr-cymru) [reply sos/plugins/juju.py]: (It is also adding, like, 12 lines to the plugin, which could otherwise be handled in a single `add_cmd_output([])` call).

File: sos/plugins/juju.py [hunk 3]
```diff
@@ -9,113 +9,201 @@
...
+                        )
+
+                    # Get the model status as json so we can dig deeper
+                    status = self.get_cmd_output_now(
+                        "sudo -i -u {} juju status -m {} --format json".format(
+                            username, modelname
+                        ),
+                        suggest_filename="juju_status_-m_{}.json".format(
+                            modelname
+                        ),
```
Comment (bmr-cymru) [inline sos/plugins/juju.py]: Same here. Why all this complexity? The whole point of mangled file names is that they fit the file system, are readable by machines, and reasonable friendly to human eyes. This is just making extra work for no real gain.

File: sos/plugins/juju.py [hunk 4]
```diff
@@ -9,113 +9,201 @@
...
+                filename = re.sub(r"_--format_yaml", "", filename)
+                filename += ".yaml"
             self.add_cmd_output(
-                "/usr/lib/juju/bin/mongoexport --ssl \
-                --dbpath=/var/lib/juju/db --db juju --collection {0} \
-                --jsonArray".format(collection),
-                suggest_filename="{}.json".format(collection))
+                "sudo -i -u {} {}".format(username, cmd),
+                suggest_filename=filename,
+            )
```
Comment (bmr-cymru) [inline sos/plugins/juju.py]: Just nack. If the standard mangling isn't working for you, then let's have a conversation about that. Driving a truck through it left right and centre in these plugins is not a maintainable or reasonable way to approach that.

File: sos/plugins/juju.py [hunk 5]
```diff
@@ -9,113 +9,201 @@
...
+        output_dir = self.get_cmd_output_path(make=False)
+        juju_config_files = glob.glob(output_dir + "/juju_config*.yaml")
+        juju_cert_files = glob.glob(output_dir + "/juju*controller*.yaml")
+
+        for file in juju_config_files:
+            self._log_debug(
+                "substituting {} for {} in {}".format(
+                    juju_config_sub_regex, juju_config_regex, file
+                )
+            )
```
Comment (bmr-cymru) [inline sos/plugins/juju.py]: Why are you doing this? The postproc hooks *already* log what they are doing:

```python
 303     def do_file_sub(self, srcpath, regexp, subst):
 311         try:
 312             path = self._get_dest_for_srcpath(srcpath)
 313             self._log_debug("substituting scrpath '%s'" % srcpath)
 314             self._log_debug("substituting '%s' for '%s' in '%s'"
 315                             % (subst, regexp, path))
```

```python
 260     def do_cmd_output_sub(self, cmd, regexp, subst):
 272         self._log_debug("substituting '%s' for '%s' in commands matching '%s'"
 273                         % (subst, regexp, globstr))
 274 
```

These are infrastructure level messages. They don't belong in individual plugins.

File: sos/plugins/juju.py [hunk 6]
```diff
@@ -9,113 +9,201 @@
...
+                result, replacements = re.subn(
+                    certs_regex, certs_sub_regex, fp.read()
+                )
+                fp.close()
+                fp = open(file, "w")
+                fp.write(result)
+                fp.close()
+            except Exception as e:
+                msg = "regex substitution failed for '%s' with: '%s'"
+                self.log_error(msg % (file, e))
```
Comment (bmr-cymru) [inline sos/plugins/juju.py]: Are you kidding me? If you don't find what you need in the canned output substitution methods provided, talk to us - we don't just throw dozens of lines of ad-hoc open code into plugins like this. It is neither sustainable nor maintainable.

File: sos/plugins/juju_controller.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+                + "--authenticationDatabase admin --ssl "
+                + "--sslAllowInvalidCertificates --username '{}' --password "
+                + "'{}' --collection {} --jsonArray"
+            )
+            self.add_cmd_output(
+                mongocmd.format(machineid, dbpass, collection),
+                suggest_filename=filename,
+            )
+
+    def get_machine_id(self):
```
Comment (bmr-cymru) [inline sos/plugins/juju_controller.py]: Not a great name, since `machine_id` now has a system wide meaning - consider `juju_machine_id`.

File: sos/plugins/juju_controller.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        )
+        self.add_cmd_output(
+            mongocmd.format(machineid, dbpass, query),
+            suggest_filename=filename,
+        )
+
+    def setup(self):
+        self.add_copy_spec("/etc/default/mongodb")
+        self.add_copy_spec("/var/lib/juju/bootstrap-params")
+        self.add_copy_spec("/lib/systemd/system/juju-db/juju-db.service")
```
Comment (bmr-cymru) [inline sos/plugins/juju_controller.py]: Don't make repeated calls to `add_copy_spec()`; it accepts either a string, or a list of strings.

File: sos/plugins/juju_controller.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+            "db.serverStatus": "JSON.stringify(db.serverStatus());",
+            "db.stats": "JSON.stringify(db.stats());",
+            "getCollectionInfos": "JSON.stringify(db.getCollectionInfos());",
+            "collections.stats": "stats = new Array(); "
+            + "db.getCollectionNames().forEach(function(coll) { "
+            + "var c = db.getCollection(coll); "
+            + "stats.push(JSON.stringify(c.stats())); "
+            + "}); "
+            + 'print("["); print(stats.join(",")); print("]")',
+        }
```
Comment (bmr-cymru) [inline sos/plugins/juju_controller.py]: This seems extremely complicated to be maintaining in an sos plugin: presumably it will need to change as Juju changes. Is there really no support tooling provided with the product that can provide the same, or similar information? That seems like a supportability gap if so...

File: sos/plugins/juju_machine.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    def setup(self):
+        # Get juju services, they are uniquely named based on the deployment
+        # so we need to parse what their names are:
+        cmd = "systemctl --no-pager --all --no-legend list-units juju\\*"
+        juju_systemctl_output = self.call_ext_prog(cmd)["output"]
+        for line in iter(juju_systemctl_output.splitlines()):
+            match = re.match(r"^(juju\S+)\s.*", line)
+            if match:
+                self.add_journal(units=str(match.group(1)))
+
```
Comment (bmr-cymru) [inline sos/plugins/juju_machine.py]: This doesn't belong in a plugin. If there is a need to obtain a list of units from `systemd`, that should be a service provided by the `Plugin` class.

File: sos/plugins/juju_machine.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        # Get agent configs for each agent
+        for dirpath, dirnames, filenames in os.walk("/var/lib/juju/agents"):
+            for dirname in dirnames:
+                self.add_copy_spec(
+                    "/var/lib/juju/agents/" + dirname + "/agent.conf"
+                )
+
```
Comment (bmr-cymru) [inline sos/plugins/juju_machine.py]: Nack. Build a list (e.g. a comprehension), and pass it to a single `add_copy_spec()` call.

File: sos/plugins/juju_machine.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        self.add_cmd_output("ls -alRh /var/log/juju*")
+        self.add_cmd_output("ls -alRh /var/lib/juju*")
```
Comment (bmr-cymru) [inline sos/plugins/juju_machine.py]: Use a list.

Comment (bmr-cymru) [reply sos/plugins/juju_machine.py]: Also, these lines appear identical.

File: sos/plugins/juju_machine.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+            # want all logs you want this too.
+            self.add_copy_spec(["/var/log/juju", "/var/lib/juju"])
+        else:
+            # We need this because we want to collect to the limit of all
+            # *.logs in the directory.
+            if os.path.isdir("/var/log/juju/"):
+                for filename in os.listdir("/var/log/juju/"):
+                    if filename.endswith(".log"):
+                        fullname = os.path.join("/var/log/juju/" + filename)
+                        self.add_copy_spec(fullname)
```
Comment (bmr-cymru) [inline sos/plugins/juju_machine.py]: What's wrong with `add_copy_spec("/var/log/juju/*.log")`? This just seems like a very complicated way to glob for *.log.

File: sos/plugins/juju_machine.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+    def apply_regex_sub(self, regexp, subst):
+        self.do_path_regex_sub("/var/lib/juju/agents/*", regexp, subst)
+
```
Comment (bmr-cymru) [inline sos/plugins/juju_machine.py]: Not sure a wrapper method is worth it for one argument substitution. Why not just put it in a local, and reference it twice:

```
agents_path = "/var/lib/juju/agents/*"
self.do_path_regex_sub(agents_path, keys_regex, sub_regex)
self.do_path_regex_sub(agents_path, certs_regex, sub_regex)
```

File: sos/policies/debian.py [hunk 15]
```diff
@@ -14,7 +14,7 @@ class DebianPolicy(LinuxPolicy):
     _debv_filter = ""
     valid_subclasses = [DebianPlugin]
     PATH = "/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games" \
-           + ":/usr/local/sbin:/usr/local/bin"
+        + ":/usr/local/sbin:/usr/local/bin:/snap/bin"
```
Comment (bmr-cymru) [inline sos/policies/debian.py]: Nack. You don't just hide a significant change to a long-standing distro policy by tucking it away at the end of pages of Juju plugin patches. This needs to be a separate commit where it can easily be found, and where it can be discussed with the Debian and Ubuntu policy maintainers.

**Comment by bmr-cymru** [comment]
This pull request **introduces 1 alert** when merging 6d9255915bb3fd5d9d9cf0083a262f45e1dc11c2 into e456ae4f23e49c36d5efe6b0584c5ec29be21ba5 - [view on LGTM.com](https://lgtm.com/projects/g/sosreport/sos/rev/pr-af11f37ada29fe9aeed5ea5ae3d2e880cd0cf21e)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by bmr-cymru** [comment]
This pull request **introduces 1 alert** when merging b9e933a37ffac8f4232474bb1fd5490c03597c56 into e456ae4f23e49c36d5efe6b0584c5ec29be21ba5 - [view on LGTM.com](https://lgtm.com/projects/g/sosreport/sos/rev/pr-bb4b66ca267e2f0fd5e4eb9b97f75282bde0c6ca)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by bmr-cymru** [comment]
This pull request **introduces 2 alerts** when merging 57c29ecfe99db537996e85c61b37f82f00bbfb3e into e456ae4f23e49c36d5efe6b0584c5ec29be21ba5 - [view on LGTM.com](https://lgtm.com/projects/g/sosreport/sos/rev/pr-4369165eaa0df34651c7cfc0d271ffa13ce47402)

**new alerts:**

* 2 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by bmr-cymru** [comment]
This pull request **introduces 2 alerts** when merging 3acac3b7f3c84d57dc97be53b916cd0228dd5acc into e456ae4f23e49c36d5efe6b0584c5ec29be21ba5 - [view on LGTM.com](https://lgtm.com/projects/g/sosreport/sos/rev/pr-8efeb1940fad04c205256ad366ce7fe46acfe09c)

**new alerts:**

* 2 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by bmr-cymru** [comment]
This pull request **introduces 2 alerts** when merging d88b7076d2286f81d2d538106b6af7e40696aa96 into e456ae4f23e49c36d5efe6b0584c5ec29be21ba5 - [view on LGTM.com](https://lgtm.com/projects/g/sosreport/sos/rev/pr-a95f9a183858ab5473df0c326b4ee53069beef4c)

**new alerts:**

* 2 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by bmr-cymru** [comment]
This PR is clearly not in a state to review: I can't find any responses to the earlier review comments (that take some time of the team to go through and write...), and now there are twelve new commits each with a subject of "Making Mods"?

That's not how we work here - please take some time to clean this branch up, put proper commit messages in place, address the review comments that were already raised, and then submit a new PR once that's done.

Everyone is getting notifications for this, but there's clearly nothing that we can work on yet.

**Comment by nniehoff** [comment]
@bmr-cymru First off, I'm very sorry for the newbie mistakes, I didn't realize making commits to my branch on my fork would update an open pull request.  This is obviously my first attempt to really contribute to another project, I'm clearly not a software engineer.  My intentions were to clean up the code with all of your original comments, squash my commits, update the pull request and respond to your comments all at once as to not bother you as much as possible.  Second, thank you very much for the review and the detailed comments I really appreciate it.  I will work on addressing all of the issues and submit new PRs per @battlemidget's request.

**Comment by battlemidget** [comment]
@nniehoff feel free to hit me up or @BryanQuigley if you want some eyes on it before you open a pr
